hazelcast/src/hazelcast/util/impl/Executor.cpp

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cmath>
+#include <boost/foreach.hpp>
+
+#include "hazelcast/util/impl/SimpleExecutorService.h"
+#include "hazelcast/client/exception/IException.h"
+#include "hazelcast/client/exception/IOException.h"
+#include "hazelcast/util/HashUtil.h"
+#include <sstream>
+
+namespace hazelcast {
+    namespace util {
+        namespace impl {
+            int32_t SimpleExecutorService::DEFAULT_EXECUTOR_QUEUE_CAPACITY = 100 * 1000;
+            util::Atomic<int64_t> SimpleExecutorService::THREAD_ID_GENERATOR;
+
+            SimpleExecutorService::SimpleExecutorService(ILogger &logger, const std::string &threadNamePrefix,
+                                                         int threadCount,
+                                                         int32_t maximumQueueCapacity)
+                    : logger(logger), threadNamePrefix(threadNamePrefix), threadCount(threadCount), live(true),
+                      workers(threadCount), maximumQueueCapacity(maximumQueueCapacity) {
+                startWorkers();
+            }
+
+            SimpleExecutorService::SimpleExecutorService(ILogger &logger, const std::string &threadNamePrefix,
+                                                         int threadCount)
+                    : logger(logger), threadNamePrefix(threadNamePrefix), threadCount(threadCount), live(true),
+                      workers(threadCount), maximumQueueCapacity(DEFAULT_EXECUTOR_QUEUE_CAPACITY) {
+                startWorkers();
+            }
+
+            void SimpleExecutorService::startWorkers() {
+                // `maximumQueueCapacity` is the given max capacity for this executor. Each worker in this executor should consume
+                // only a portion of that capacity. Otherwise we will have `threadCount * maximumQueueCapacity` instead of
+                // `maximumQueueCapacity`.
+                int32_t perThreadMaxQueueCapacity = static_cast<int32_t>(ceil(
+                        (double) 1.0 * maximumQueueCapacity / threadCount));
+                for (int i = 0; i < threadCount; i++) {
+                    workers[i].reset(new Worker(threadNamePrefix, perThreadMaxQueueCapacity, live, logger));
+                    workers[i]->start();
```

Comment:
From the current usage, it looks like there is no need for StartedThread and Thread differentiation. It looks like we can use old Thread class without any change because constructing a thread with is always together with starting it.  
What is the reason for introducing `StartedThread`?

Reply:
For some of the spi services, there is a start method where I start a thread but I construct the thread at the service constructor.

My executor implementation is missing the management of workers as being done at Java where the workers are being started on need basis. I want to also change it to this version later on.

We can always achieve the same purpose by using smart pointers and with some tricks but I find it better usage starting the the thread at another call which is the way chosen by Java. 

At this PR we provide both ways of usage, hence it should not be a problem. Name StartedThread implies better that the thread is started by construction.